### PR TITLE
Fix adding multiple event handlers with same function body

### PR DESCRIPTION
**Pull request by bastuijnman**
Hello,

When trying to add multiple event listeners to a widget, with the same function body, only the first event gets fired. This is because the events are stored on an object with the function as a key. Because javascript internally does `.toString` on the function when applying it as key it can only store one event with that function body.

Normally this is not a problem, but when using `.bind` the function body will automatically become `function () { [native code] }` and thus only one bound function will be able to be set to that listener.

This fix checks for the function reference itself and thus is able to add multiple bound event listeners.

### CODE REVIEW

**Comment by sdclibbery** [comment]
Hi, this looks great :-)

We are aware of this problem, and its fantastic to have a fix for it :-)

Previously, if a caller was silly enough to call addEventListener several times with the exact same function, then that function would still only get called back once for each event. However I think that after this Pull Request, the function would get called once for each time it had been added (because each time addEventListener is called, a new reference to the function will be pushed onto the array).

I wonder if we need to add a check in addEventListener to see whether the function already exists in the array, and if so not add it again? I think this would help give us confidence that existing (possibly buggy) code would not get broken by this change :-)

Thanks,
Steve

**Comment by bastuijnman** [comment]
I added a check for the function reference before adding it, this should solve the problem with buggy code being executed twice ;)

**Comment by bastuijnman** [comment]
My colleague reminded me that the same fix also needed to be applied to the meta-events, so I added that as well

**Comment by bastuijnman** [comment]
Any status update?

**Comment by sdclibbery** [comment]
Hi, we will be looking at this pull request on Wednesday afternoon and we should be able to merge it then.

**Comment by bastuijnman** [comment]
Cool thanks!

**Comment by tsadlerBBC** [comment]
Hi Bas,

We're currently testing your changes and are having a problem with one of our apps. We'll continue working on it, and we still intend to merge this in once we've determined the cause of the problem.

Thanks,

Tom

**Comment by bastuijnman** [comment]
Hey Tom,

Sounds allright, if you need any info I'll be glad to help :)

File: static/script/events/event.js [hunk 1]
```diff
@@ -115,10 +116,11 @@ require.def('antie/events/event',
 			 * @param {Function} func The handler to be removed.
 			 */
 			removeEventListener: function(ev, func) {
```
Comment (tsadlerBBC) [inline static/script/events/event.js]: Hi Bas,

I've got to the bottom of the problem. Our app is calling removeEventListener where no event has been added.

We will be fixing this in our app, but could you also please add some error handling around this? We would expect a guard statement at the top of this function along the lines of:

```javascript
                if (!this._eventListeners[ev]) {
                    RuntimeContext.getDevice().getLogger().error("Attempting to remove non-existent event listener");
                    return;
                }
```

Thanks,

Tom

**Comment by bastuijnman** [comment]
Hi Tom,

That's a bit of a weird case :smile: but I added the check + an extra test for the widget.js

Kind regards,

Bas Tuijnman - 24i

**Comment by tsadlerBBC** [comment]
It is indeed a weird case! Just worried in case we upgrade the app to a version of TAL with this in before we fix the bug :smile:.

Thanks again for this. As Stephen said above, we wanted a fix for this issue so it's great that we've got this now.

Tom

**Pull request merged**